apps/condo/domains/billing/schema/ValidateQRCodeService.js

```diff
@@ -0,0 +1,128 @@
+        type: WRONG_FORMAT,
+        message: 'Provided QR code doesn\'t have all required fields',
+    },
+    NO_ORGANIZATION: {
+        mutation: 'validateQRCode',
+        code: INTERNAL_ERROR,
+        type: NOT_FOUND,
+        message: 'Organization with provided TIN is not registered with Doma.ai',
+    },
+    NO_ACQUIRING_CONTEXT: {
+        mutation: 'validateQRCode',
+        code: INTERNAL_ERROR,
+        type: NOT_FOUND,
+        message: 'Organization with provided TIN does not have an active AcquiringIntegrationContext',
+    },
+}
+
+/**
+ * @typedef {object} ValidateQRCodeInput
+ * @property {string} qrCode
+ */
+
+/**
+ * @typedef {object} ValidateQRCodeInputArgs
+ * @property {ValidateQRCodeInput} data
+ */
+
+const ValidateQRCodeService = new GQLCustomSchema('ValidateQRCodeService', {
+    types: [
+        {
+            access: true,
+            type: 'input ValidateQRCodeInput { dv: Int!, sender: SenderFieldInput!, qrCode: String! }',
+        },
+        {
+            access: true,
+            type: 'type ValidateQRCodeOutput { qrCodeFields: JSON! }',
+        },
+    ],
+
+    mutations: [
+        {
+            access: access.canValidateQRCode,
+            schema: 'validateQRCode(data: ValidateQRCodeInput!): ValidateQRCodeOutput',
+            resolver: async (parent, /** @type {ValidateQRCodeInputArgs} */ args, context) => {
+                const { data } = args
+                const { qrCode } = data
+
+                const checkLimits = async (ip) => {
+                    await redisGuard.checkCustomLimitCounters(
+                        `validate-QR-code-${ip}`,
+                        BILLING_VALIDATE_QR_CODE_WINDOW,
+                        MAX_CLIENT_VALIDATE_QR_CODE_BY_WINDOW,
+                    )
+                }
+
+                const ip = context.req.headers['x-forwarded-for'] || context.req.socket.remoteAddress
+                await checkLimits(ip)
+
+                const matches = /^ST(?<version>\d{4})(?<encodingTag>\d)\|(?<requisitesStr>.*)$/g.exec(qrCode)
+                const encodingTag = get(matches, ['groups', 'encodingTag'])
+                const requisitesStr = get(matches, ['groups', 'requisitesStr'], '')
+
+                // TODO(AleX83Xpert): maybe decode requisitesStr
+                // Need to test the result of scanning from mobile devices
+                // https://encoding.spec.whatwg.org/#koi8-r
+                // https://encoding.spec.whatwg.org/#windows-1251
+                // const encoding = get(['windows-1251', 'utf-8', 'koi8-r'], encodingTag, 'utf-8')
+
+                const qrCodeFields = Object.fromEntries(requisitesStr.split('|').map(part => part.split('=', 2)))
+                const { PersonalAcc, PersAcc, BIC, PayerAddress, LastName, PaymPeriod, Sum, PayeeINN } = qrCodeFields
+
+                if (!BIC || !PayerAddress || !LastName || !PaymPeriod || !Sum || !PersAcc || !PayeeINN || !PersonalAcc) throw new GQLError(ERRORS.INVALID_QR_CODE, context)
```

Comment:
Better to move required fields to constant

Reply:
Refactored